suite/tests/client-interface/drbbdup-emul-reg-clobber-test.dll.c

```diff
@@ -0,0 +1,311 @@
+    CHECK(res == DRBBDUP_SUCCESS, "failed to register case 2");
+
+    *enable_dups = true;
+    *enable_dynamic_handling = false; /* disable dynamic handling */
+
+    return 0; /* return default case */
+}
+
+static void
+analyze_case(void *drcontext, void *tag, instrlist_t *bb, uintptr_t encoding,
+             void *user_data, void *orig_analysis_data, void **analysis_data)
+{
+    per_block_t *per_block =
+        (per_block_t *)dr_thread_alloc(drcontext, sizeof(per_block_t));
+    memset(per_block, 0, sizeof(*per_block));
+    *analysis_data = per_block;
+}
+
+static void
+destroy_case_analysis(void *drcontext, uintptr_t encoding, void *user_data,
+                      void *orig_analysis_data, void *analysis_data)
+{
+    per_block_t *per_block = (per_block_t *)analysis_data;
+    CHECK(per_block->saw_first, "failed to see first instr");
+    /* If we added a rest-of-block emul it will hide the last instr. */
+    CHECK(per_block->saw_last || has_rest_of_block_emul, "failed to see last instr");
+    dr_thread_free(drcontext, analysis_data, sizeof(per_block_t));
+}
+
+static void
+write_aflags(void *drcontext, instrlist_t *bb, instr_t *inst)
+{
+    instrlist_meta_preinsert(bb, inst,
+                             XINST_CREATE_cmp(drcontext, opnd_create_reg(DR_REG_START_32),
+                                              OPND_CREATE_INT32(0)));
+    instrlist_meta_preinsert(bb, inst,
+                             XINST_CREATE_cmp(drcontext, opnd_create_reg(DR_REG_START_32),
+                                              OPND_CREATE_INT32(-1)));
+}
+
+static dr_emit_flags_t
+instrument_instr(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                 instr_t *where, bool for_trace, bool translating, uintptr_t encoding,
+                 void *user_data, void *orig_analysis_data, void *analysis_data)
+{
+    bool is_first, is_last;
+    drbbdup_status_t res;
+    per_block_t *per_block = (per_block_t *)analysis_data;
+
+    res = drbbdup_is_first_instr(drcontext, instr, &is_first);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to check whether instr is first");
+    if (is_first) {
+        per_block->saw_first = true;
+        /* Ensure DR_EMULATE_REST_OF_BLOCK didn't leak through. */
+        const emulated_instr_t *emul_info;
+        if (drmgr_in_emulation_region(drcontext, &emul_info)) {
+            /* It might be the rep string, which sets DR_EMULATE_INSTR_ONLY.
+             * XXX: We could try to pass a flag: but drbbdup won't let
+             * us pass from app2app; we'd need TLS or a global.
+             */
+            CHECK(!TEST(DR_EMULATE_REST_OF_BLOCK, emul_info->flags) ||
+                      TEST(DR_EMULATE_INSTR_ONLY, emul_info->flags),
+                  "DR_EMULATE_REST_OF_BLOCK leaked through!");
+        }
+    }
+
+#ifdef X86
+    /* Cause drutil's expanded "rep movs" to crash if drbbdup sets
+     * DRREG_IGNORE_CONTROL_FLOW (i#5398).  We do not spill rcx until after the
```

Comment:
This is testing something else and can also be removed, right?

Reply:
Removed.